src/katello/repos.py

```diff
@@ -0,0 +1,93 @@
+import errno
+import os
+import sys
+
+from katello.rhsm import get_uep, lookup_consumer_id
+
+from rhsm.connection import GoneException, RemoteServerException
+
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+def error_message(msg):
+    sys.stderr.write(msg + "\n")
+
+def report_enabled_repos(consumer_id, report):
+    """
+    Report enabled repositories to the UEP.
+    :param consumer_id: The consumer ID.
+    :type consumer_id: str
+    :param report: The report to send.
+    :type report: dict
+    """
+    uep = get_uep()
+    method = '/systems/%s/enabled_repos' % uep.sanitize(consumer_id)
+    try:
+        uep.conn.request_put(method, report)
+    except (RemoteServerException, GoneException):
+        error = sys.exc_info()[1] # backward and forward compatible way to get the exception
+        error_message(str(error))
+
+def upload_enabled_repos_report(report):
+    content = report.content
+    consumer_id = lookup_consumer_id()
+    if consumer_id is None:
+        error_message('Cannot upload enabled repos report, is this client registered?')
+    else:
+        cache = EnabledRepoCache(consumer_id, content)
+        if not cache.is_valid():
+            report_enabled_repos(consumer_id, content)
+            cache.save()
+
+
+class EnabledRepoCache:
+    CACHE_FILE = '/var/cache/katello-agent/enabled_repos.json'
+
+    def __init__(self, consumer_id, content):
+        self.consumer_id = consumer_id
+        self.content = content
+
+    @staticmethod
+    def remove_cache():
+        try:
+            os.remove(EnabledRepoCache.CACHE_FILE)
+        except OSError:
+            pass
+
+    def is_valid(self):
+        cache_file = None
+        # wrapper try block w/ finally needed for python 2.4
+        try:
+            try:
+                cache_file = open(self.CACHE_FILE, 'r')
+                try:
+                    return self.data() == json.loads(cache_file.read())
+                except ValueError:
+                    return False
+            except OSError:
+                error = sys.exc_info()[1] # backward and forward compatible way to get the exception
+                if error.errno == errno.ENOENT:
+                    return False
+        finally:
```

Comment:
Can't this finally block be part of the try/except so you avoid the double indentation?

Reply:
Apparently python 2.4 on EL5 does not support try/except/finally, but only try/finally or try/except

Reply:
I'd forgotten what a pain python 2.4 was.